### PR TITLE
Feature/55a recipes view

### PR DESCRIPTION
**Pull request by Munin33**
this branch adds the recipe and recipe execution UIs

### CODE REVIEW

File: src/utils/Utils.ts [hunk 1]
```diff
@@ -131,6 +139,55 @@ export default class Utils {
...
+    return executions.map((execution) => {
+      return {
+        id: execution.id || execution.identifier,
+        firstValue: 'Execution Identifier: ' + execution.identifier,
+        secondValue: Utils.getExecutionStateString(execution),
+        thirdValue: Utils.getExecutionProgress(execution),
+        link: onClick ? {
+          onClick: () => { onClick(execution.id || execution.identifier); },
+        } : {
+            link: { name: 'execution', params: { id: execution.id } },
```
Comment (friedow) [inline src/utils/Utils.ts]: Intendation is broken.

File: src/utils/Utils.ts [hunk 2]
```diff
@@ -131,6 +139,55 @@ export default class Utils {
...
+    return execution.processingStates.map((state) => {
+      return {
+        id: state.identifier,
+        firstValue: 'State identifier: ' + state.identifier,
+        secondValue: Utils.getExecutionStateString(state),
+      };
+    });
+  }
+
+  public static getExecutionStateString(execution: OptimizationExecutionObject): string {
```
Comment (friedow) [inline src/utils/Utils.ts:170]: Please use guard clauses in this function and a default return at the end.

```ts
if (execution.finishedAt) {
  if (execution.successful) {
    return `Finished successful ${ta.ago(execution.finishedAt)}`;
  } else {
    return `Execution failed ${ta.ago(execution.finishedAt)}`;
  }
} else if (execution.startedAt) {
  return `Started ${ta.ago(execution.startedAt)}`;
}

return 'Execution was not started.';
```

File: src/utils/Utils.ts [hunk 3]
```diff
@@ -131,6 +139,55 @@ export default class Utils {
...
+      executionStateString = `Started ${ta.ago(execution.startedAt)}`;
+    }
+
+    return executionStateString;
+  }
+
+  public static getExecutionProgress(execution: OptimizationExecution): string {
+    const numberOfIngredients = execution.processingStates.length;
+    const numberOfFinishedIngredients = execution.processingStates.filter((state) => state.successful).length;
+    const finishedPercentage = (numberOfFinishedIngredients / numberOfIngredients) * 100;
```
Comment (friedow) [inline src/utils/Utils.ts:206]: Parentheses are unnecessary here.

File: src/utils/Utils.ts [hunk 4]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+  public static async recipeAttributesToList(recipe: Recipe): Promise<ListEntry[]> {
+    return [
+      {
+        id: 'name',
+        firstValue: 'Name',
+        secondValue: recipe.name,
+      },
+      {
+        id: 'Input Ingrediants',
+        firstValue: 'Input Ingrediants',
```
Comment (friedow) [inline src/utils/Utils.ts]: Typo: IngrediEnts

File: src/utils/Utils.ts [hunk 5]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+      },
+      {
+        id: 'outputIngredients',
+        firstValue: 'Output Ingrediants',
+        secondValue: recipe.ingredients ? await this.getOutputIngrediantList(recipe.ingredients) : '',
+      },
+    ];
+  }
+
+  public static async getInputIngrediantList(ingredients: Ingredient[]) {
```
Comment (friedow) [inline src/utils/Utils.ts]: Typo: IngrediEnts

File: src/utils/Utils.ts [hunk 6]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+        } else {
+          inputList = inputList + ', ' + resourceType.name;
+        }
+      });
+    };
+    await getNames();
+    return inputList;
+  }
+
+  public static async getTransformerIngrediantList(ingredients: Ingredient[]) {
```
Comment (friedow) [inline src/utils/Utils.ts]: Typo: IngrediEnts

File: src/utils/Utils.ts [hunk 7]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+        secondValue: recipe.name,
+      },
+      {
+        id: 'Input Ingrediants',
+        firstValue: 'Input Ingrediants',
+        secondValue: recipe.ingredients ? await this.getInputIngrediantList(recipe.ingredients) : '',
+      },
+      {
+        id: 'transformerIngredients',
+        firstValue: 'Transformer Ingrediants',
```
Comment (friedow) [inline src/utils/Utils.ts]: Typo: IngrediEnts

File: src/utils/Utils.ts [hunk 8]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+        secondValue: recipe.ingredients ? await this.getTransformerIngrediantList(recipe.ingredients) : '',
+      },
+      {
+        id: 'algorithmIngredient',
+        firstValue: 'Algorithm Ingrediants',
```
Comment (friedow) [inline src/utils/Utils.ts]: Typo: IngrediEnts

File: src/utils/Utils.ts [hunk 9]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+        secondValue: recipe.ingredients ? await this.getAlgorithmIngrediantList(recipe.ingredients) : '',
+      },
+      {
+        id: 'outputIngredients',
+        firstValue: 'Output Ingrediants',
```
Comment (friedow) [inline src/utils/Utils.ts]: Typo: IngrediEnts

File: src/utils/Utils.ts [hunk 10]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+        } else {
+          transformerList = transformerList + ', ' + transformer.name;
+        }
+      });
+    };
+    await getNames();
+    return transformerList;
+  }
+
+  public static async getAlgorithmIngrediantList(ingredients: Ingredient[]) {
```
Comment (friedow) [inline src/utils/Utils.ts]: Typo: IngrediEnts

File: src/utils/Utils.ts [hunk 11]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+        } else {
+          algorithmList = algorithmList + ', ' + algorithm.name;
+        }
+      });
+    };
+    await getNames();
+    return algorithmList;
+  }
+
+  public static async getOutputIngrediantList(ingredients: Ingredient[]) {
```
Comment (friedow) [inline src/utils/Utils.ts]: Typo: IngrediEnts

File: src/utils/Utils.ts [hunk 12]
```diff
@@ -131,6 +139,55 @@ export default class Utils {
     ];
   }
 
+
+  public static optimizationExecutionsToList(executions: OptimizationExecution[], onClick?: clickHandler) {
```
Comment (friedow) [inline src/utils/Utils.ts]: Missing return type annotation.

File: src/utils/Utils.ts [hunk 13]
```diff
@@ -172,6 +229,20 @@ export default class Utils {
     ];
   }
 
+  public static recipesToList(recipes: Recipe[], onClick?: clickHandler) {
```
Comment (friedow) [inline src/utils/Utils.ts]: Missing return type annotation.

File: src/utils/Utils.ts [hunk 14]
```diff
@@ -131,6 +139,55 @@ export default class Utils {
...
+  }
+
+  public static getExecutionProgress(execution: OptimizationExecution): string {
+    const numberOfIngredients = execution.processingStates.length;
+    const numberOfFinishedIngredients = execution.processingStates.filter((state) => state.successful).length;
+    const finishedPercentage = (numberOfFinishedIngredients / numberOfIngredients) * 100;
+    return `${Math.round(finishedPercentage)} %`;
+  }
+
   public static transformerToList(transformers: Transformer[], onClick?: clickHandler) {
```
Comment (friedow) [inline src/utils/Utils.ts]: Missing return type annotation.

File: src/utils/Utils.ts (hunk 5 above)
Comment (friedow) [inline src/utils/Utils.ts]: Missing return type annotation.

File: src/utils/Utils.ts (hunk 6 above)
Comment (friedow) [inline src/utils/Utils.ts]: Missing return type annotation.

File: src/utils/Utils.ts (hunk 10 above)
Comment (friedow) [inline src/utils/Utils.ts]: Missing return type annotation.

File: src/utils/Utils.ts (hunk 11 above)
Comment (friedow) [inline src/utils/Utils.ts]: Missing return type annotation.

File: src/utils/Utils.ts [hunk 15]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+        } else {
+          outputList = outputList + ', ' + resourceType.name;
+        }
+      });
+    };
+    await getNames();
+    return outputList;
+  }
+
+  public static async asyncForEach(array: any, callback: any) {
```
Comment (friedow) [inline src/utils/Utils.ts]: Missing return type annotation.

File: src/utils/Utils.ts [hunk 16]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+  public static async getInputIngrediantList(ingredients: Ingredient[]) {
+    const inputIngredients = ingredients.filter( (ingredient) => {
+      return ingredient.ingredientType === IngredientType.INPUT;
+    });
+    let inputList = '';
+    const getNames = async () => {
+      await this.asyncForEach(inputIngredients, async (ingredient: any) => {
+        const resourceType = await ResourceTypes.getOne(ingredient.ingredientDefinition);
+        if (inputList === '') {
+        inputList = resourceType.name;
```
Comment (friedow) [inline src/utils/Utils.ts]: Intendation broken.

File: src/utils/Utils.ts [hunk 17]
```diff
@@ -194,6 +265,118 @@ export default class Utils {
...
+      },
+    ];
+  }
+
+  public static async getInputIngrediantList(ingredients: Ingredient[]) {
+    const inputIngredients = ingredients.filter( (ingredient) => {
+      return ingredient.ingredientType === IngredientType.INPUT;
+    });
+    let inputList = '';
+    const getNames = async () => {
```
Comment (friedow) [inline src/utils/Utils.ts]: There is a nice construct for that called Promise.all( iterable: Promise[] ). Here is the documentation for that: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Promise/all.

This is what you want to do here:
```ts
const namePromises = inputIngredients.map(async (ingredient: Ingredient): Promise<string> => {
  const resourceType = await ResourceTypes.getOne(ingredient.ingredientDefinition!);
  return resourceType.name;
});
const names = await Promise.all(namePromises);
return names.join(', ');
```

This way your code is not only shorter, but it also executes in a true-async behavior (without waiting in for loops). Please adjust the other util functions below accordingly to this example.

File: src/views/Execution.vue [hunk 18]
```diff
@@ -0,0 +1,74 @@
+<template>
+  <main>
+    <ViewHeader :title="execution.identifier" :backLink="{ link: { name: 'executions' } }"/>
+
+    <ListSection class="preview-container" :list="executionList"/>
+
+    <div class="row">
```
Comment (friedow) [inline src/views/Execution.vue]: Unnecessary div.

File: src/views/Recipe.vue [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+    recipeActions.push(
+      {
+        id: '1',
+        firstValue: 'Create a Recipe',
+        link: {
+          link: { name: 'create-recipe'},
+        },
+      },
+      {
+      id: '2',
```
Comment (friedow) [inline src/views/Recipe.vue]: Intendation broken.

**Review by friedow — APPROVED** [review]

File: src/utils/Utils.ts (hunk 1 above)
Comment (Munin33) [reply src/utils/Utils.ts]: fixed it

File: src/utils/Utils.ts (hunk 4 above)
Comment (Munin33) [reply src/utils/Utils.ts]: jup. more than once.

File: src/utils/Utils.ts (hunk 15 above)
Comment (Munin33) [reply src/utils/Utils.ts]: added it.

File: src/utils/Utils.ts (hunk 16 above)
Comment (Munin33) [reply src/utils/Utils.ts]: not anymore!

File: src/utils/Utils.ts (hunk 17 above)
Comment (Munin33) [reply src/utils/Utils.ts]: I really liked the asyncForEach, but this is better.

File: src/views/Recipe.vue (hunk 19 above)
Comment (Munin33) [reply src/views/Recipe.vue]: fixed it

File: src/utils/Utils.ts (hunk 17 above)
Comment (Munin33) [reply src/utils/Utils.ts]: thanks for pointing this out!

File: src/utils/Utils.ts (hunk 2 above)
Comment (Munin33) [reply src/utils/Utils.ts:170]: done

File: src/views/Execution.vue (hunk 18 above)
Comment (Munin33) [reply src/views/Execution.vue]: deleted

**Pull request merged**